src/Fantomas.Tests/PreserveEOLTests.fs

```diff
@@ -0,0 +1,462 @@
+      /// This includes all interactively compiled code, including #load, definitions, and expressions
+      isInteractive : bool;
+      // Indicates the code generated is an interactive 'it' expression. We generate a setter to allow clearing of the underlying
+      // storage, even though 'it' is not logically mutable
+      isInteractiveItExpr : bool;
+      // Indicates System.SerializableAttribute is available in the target framework
+      netFxHasSerializableAttribute : bool;
+      /// Whenever possible, use callvirt instead of call
+      alwaysCallVirt : bool }
+"""
+
+[<Test>]
+let ``should keep well-aligned comments``() =
+    formatSourceString false """
+/// XML COMMENT
+// Other comment
+let f() = 
+    // COMMENT A
+    let y = 1
+    (* COMMENT B *)
+    (* COMMENT C *)
+    x + x + x
+"""  config
+    |> should equal """
+/// XML COMMENT
+// Other comment
+let f() =
+    // COMMENT A
+    let y = 1
+    (* COMMENT B *)
+    (* COMMENT C *)
+    x + x + x
+"""
+
+[<Test>]
+let ``active patterns``() =
+    formatSourceString false """
+let (|Even|Odd| ) input = if input % 2=0 then Even else Odd
+
+let (|Integer|_|) (str:string) =
+    let mutable intvalue = 0
+    if System.Int32.TryParse(str, &intvalue) then Some(intvalue)
+    else None
+
+let (|ParseRegex|_|) regex str =
+    let m = Regex(regex).Match(str)
+    if m.Success
+    then Some (List.tail [ for x in m.Groups -> x.Value ])
+    else None""" config
+    |> should equal """
+let (|Even|Odd|) input = if input % 2 = 0 then Even else Odd
+
+let (|Integer|_|) (str : string) =
+    let mutable intvalue = 0
+    if System.Int32.TryParse (str, &intvalue) then Some (intvalue)
+    else None
+
+let (|ParseRegex|_|) regex str =
+    let m = Regex(regex) .Match(str)
+    if m.Success
+    then Some (List.tail [ for x in m.Groups -> x.Value ])
+    else None
+"""
+
+[<Test>]
+let ``should keep the attribute on top of the function``() =
+    formatSourceString false """[<ExtensionA>]
+type Funcs = 
+    [<ExtensionB>]
+    static member ToFunc (f: Action<_,_,_>) =
+        Func<_,_,_,_>(fun a b c -> f.Invoke(a,b,c))
+    """ config
+    |> should equal """[<ExtensionA>]
+type Funcs =
+    [<ExtensionB>]
+    static member ToFunc(f : Action<_, _, _>) =
+        Func<_, _, _, _> (fun a b c -> f.Invoke (a, b, c))
+"""
+
+[<Test>]
+let ``attributes on expressions``() =
+    formatSourceString false """
+    [<XYZ>]  
+    let   x="1"
+    let [<Literal>]y="2" 
+    [<Dependency("FSharp.Compiler", LoadHint.Always)>]
+    do ()""" config
+    |> should equal """
+    [<XYZ>]
+    let x = "1"
+    let [<Literal>] y = "2"
+    [<Dependency("FSharp.Compiler", LoadHint.Always)>]
+    do ()
+"""
+
+[<Test>]
+let ``array values``() =
+    formatSourceString false """
+let x = [ 1
+          2]
+let y = [| 3;4
+           5;6|]
+let z = [7; 8]
+    """ { config with SemicolonAtEndOfLine = false }
+    |> should equal """
+let x = [ 1;
```

Comment:
Just curious: why is this adding a `;` when `SemicolonAtEndOfLine = false` is set?  I guess it is unrelated to this PR